flink-runtime/src/main/java/org/apache/flink/runtime/state/OperatorStateRestoreOperation.java

```diff
@@ -168,9 +170,25 @@ public Void restore() throws Exception {
                     }
                 }
 
+                // sort state by offsets
+                List<Map.Entry<String, OperatorStateHandle.StateMetaInfo>> entries =
+                        stateHandle.getStateNameToPartitionOffsets().entrySet().stream()
+                                .sorted(
+                                        Comparator.comparingLong(
+                                                entry -> {
+                                                    OperatorStateHandle.StateMetaInfo
+                                                            stateMetaInfo = entry.getValue();
+                                                    long[] offsets = stateMetaInfo.getOffsets();
+                                                    if (offsets == null || offsets.length == 0) {
+                                                        return Long.MIN_VALUE;
+                                                    } else {
+                                                        return offsets[0];
+                                                    }
```

Comment:
```suggestion
   Preconditions.checkNotNull(offsets);
   Preconditions.checkState(offsets.length > 0);
   return offsets[0];                                                                        
```

I think we should throw exceptions as early as possible.

Reply:
It's confusing for me too, but we do encountered a few states with empty offset. The operator states deserialization code also ignores states with empty offset(https://github.com/apache/flink/blob/7cae20f22a24403210ca5d6addf83571fcf46843/flink-runtime/src/main/java/org/apache/flink/runtime/state/OperatorStateRestoreOperation.java#L225). I think throwing exception here might break backward compatibility.

Reply:
Thanks for the clarification.
According to the [constructor](https://github.com/apache/flink/blob/master/flink-runtime/src/main/java/org/apache/flink/runtime/state/OperatorStateHandle.java#L63C17-L63C17)  of `StateMetaInfo` and the [serde](https://github.com/apache/flink/blob/master/flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/metadata/MetadataV2V3SerializerBase.java#L626) of `OperatorStateHandle`, I think the `offsets` can not be null.  And let it stay as it is. :)

Reply:
Yeah I think so. The case we encountered should be a zero-length array.